### ISSUE TITLE
Ambiguous Reason Code in Disconnection Message

### ISSUE THREAD

**Issue by jwrct** [tronprotocol/java-tron#5989]
#### Software Versions

```
OS : Linux
JVM : Oracle Corporation 1.8.0_161 amd64
Git : a8ad2a169e58946b5b8de6ecf7b7ef5b8db05aff
Version : 4.7.5
Code : 18306
```

#### Expected behaviour
When disconnected, the exact reason for the disconnection can be obtained through the disconnection message.

#### Actual behaviour
When running Java-tron, I found that if I wanted to disconnect a peer in certain scenarios, the disconnect message sent carried an ambiguous reason code. You can take a look at the example below.

Example:
The node receives the block sent by the peer and then disconnects the peer due to block signature verification failure, but the reason for the disconnection message being sent is UNKNOWN.
```
10:17:39.014 WARN  [peerWorker-1] [net](P2pEventHandlerImpl.java:258) Message from /172.31.0.220:10850 process failed, type: BLOCK Num:58698013,ID:00000000037fa91d2bbac58ad77bd462a2878178e730eb93a4d3e00e21cbe52e, trx size: 4 type: (15, block sign error), detail: valid signature failed.

10:17:39.014 INFO  [peerWorker-1] [net](PeerConnection.java:175) Send peer /172.31.0.220:10850 message type: P2P_DISCONNECT reason: UNKNOWN
```
Therefore, after receiving the disconnection message, the peer cannot accurately know the specific reason for being disconnected. I think the reason code carried in the disconnect message in this example can be clearer. In addition, do you think it is necessary to check whether there are other disconnection scenarios with similar problems?

#### Frequency
It occurs every time in the above scenarios.

#### Steps to reproduce the behaviour
Send a block with an invalid signature to the node.

#### Backtrace

````
[backtrace]
````

When submitting logs: please submit them as text and not screenshots.

**Comment by abn2357** [comment]
Except block signature verification failure, is there any other usual peer disconnection scenario?

**Comment by jwrct** [comment]
@abn2357 There are many scenarios where the connection with the peer will be disconnected, such as: handshake failure, data acquisition timeout, receiving messages sent by the peer that do not conform to the protocol, etc.

**Comment by fyyhtx** [comment]
Are there any other disconnect scenarios that have an ambiguous reason code?

**Comment by xxo1shine** [comment]
@jwrct  I see that BAD_PROTOCOL is used in many scenarios. When I receive a disconnection reason of BAD_PROTOCOL, how can I know which scenario the peer has triggered? Do you need to solve this problem?

**Comment by fyyhtx** [comment]
@xxo1shine Can you specifically list the scenarios where BAD_PROTOCOL would be used? Also, I think you might want to consider whether it's appropriate to define this as a problem, since the original design was to carry the BAD_PROTOCOL reason code for all behaviors that do not comply with the protocol and lead to disconnection.

**Comment by jwrct** [comment]
@xxo1shine Can you specifically list the scenarios where BAD_PROTOCOL would be used? Also, I think you might want to consider whether it's appropriate to define this as a problem, since the original design was to carry the BAD_PROTOCOL reason code for all behaviors that do not comply with the protocol and lead to disconnection.

**Comment by fyyhtx** [comment]
@jwrct Have there been any new discoveries since the last reply?

**Comment by jwrct** [comment]
@fyyhtx Apart from the two scenarios mentioned earlier, no new ambiguous reason codes have been found so far.

**Comment by 317787106** [comment]
So, how to make disconnected message more clear? give more detailed reason behind UNKNOWN ?

**Comment by xxo1shine** [comment]
@jwrct  For example, `BAD_PROTOCOL` is used in both places below. 
```
one.ifPresent(peer -> disconnectFromPeer(peer, ReasonCode.BAD_PROTOCOL,
          DisconnectCause.ISOLATE2_ACTIVE));

one.ifPresent(peer -> disconnectFromPeer(peer, ReasonCode.BAD_PROTOCOL,
          DisconnectCause.ISOLATE2_ACTIVE));
```



When I receive a disconnect message from node A with reason code `BAD_PROTOCOL`, I don’t know which logic caused the problem on the other node. 

I just want to know whether this issue is going to solve this problem, or whether there are plans to solve this problem in the future, or whether this problem needs to be solved.

**Comment by jwrct** [comment]
@xxo1shine I understand that you mean to refine the scenarios of BAD_PROTOCOL and then provide more specific reason codes based on different scenarios, right? I think this is worth discussing, and we can ask for more user suggestions.

**Comment by jwrct** [comment]
@char23-web What is your point? The code you commented on has been mentioned by @xxo1shine before. Please be more specific.

**Comment by Jake151987** [comment]
@char23-web, if you have no more comments or ideas involving this issue, it would be closed for now.

**Comment by fyyhtx** [comment]
@jwrct I discovered a scenario where the reason code is unreasonable. When a received P2P_HELLO message carries a block ID whose length is not equal to Sha256Hash.LENGTH (32), the P2P_DISCONNECT message sent during exception handling carries the reason code as UNKNOWN. Is this the actual situation?

**Comment by jwrct** [comment]
@fyyhtx You are right. When the length of the blockId carried by the P2P_HELLO message is not 32, the following code will report an exception:
```
public Sha256Hash(long num, byte[] hash) {
    byte[] rawHashBytes = this.generateBlockId(num, hash);
    checkArgument(rawHashBytes.length == LENGTH);
    this.bytes = rawHashBytes;
  }

public static void checkArgument(boolean expression) {
    if (!expression) {
      throw new IllegalArgumentException();
    }
  }
```
When handling exceptions, IllegalArgumentException will be treated as an unknown exception, so the reason code carried when sending the P2P_DISCONNECT message is UNKNOWN.
This scenario can be optimized. At the beginning of processing the P2P_HELLO message, check whether the length of the carried blockId is 32. If not, carry the reason code INCOMPATIBLE_PROTOCOL when sending the P2P_DISCONNECT message.

**Comment by jwrct** [comment]
**Milestone Update (2025-05-16)**  
- **Current Status**: `In Development` 
- **Next Steps**:  
  - Full-chain testing expected to be completed by 2025-05-23  
  - Submit code for review by 2025-05-26  
- **Responsible Persons**: @jwrct
